blog/_posts/2017-11-01-zinc-blog-1.0.md

```diff
@@ -0,0 +1,258 @@
+[knirski][], [Thierry Treyer][], [Grzegorz Kossakowski][], [Dale Wijnand][] and
+[Eugene Yokota][].
+
+I'm happy that I and [Martin Duhem] met some of these contributors during our
+**[Scala Center sprees](https://github.com/scalacenter/sprees)** all over Europe
+and that they dared to submit pull requests to Zinc; while it may at first sound
+daunting, your work has had a clear impact in a piece of software widely used by
+Scala developers.
+
+### A preview of Zinc 1.0: an example of a 7x speedup
+  
+How big are the improvements that Zinc 1.x brings to the table and how do they
+impact your day-to-day developer workflow?
+
+To answer this question, we're going to use both Zinc 0.13.x and 1.x to compile
+a big real-world project: [Scalatest](http://www.scalatest.org), with
+**405.084** lines of Scala code without counting comments and blank lines.
+
+We illustrate our example with a common and relatively uninstrusive operation in
+every codebase: a method addition. In our case, we add a new method to a widely
+dependent class like `AndHaveWord` (in `Matcher.scala`) after we've correctly
+warmed up the compiler and compiled the project. How fast can Zinc 0.13.x
+recompile the new method?
+  
+[![scastie](/resources/img/blog/zinc-0.13-scalatest.gif)](/resources/img/blog/zinc-0.13-scalatest.gif)
+  
+The incremental compilation in **Zinc 0.13.x yields a compilation time of 21
+seconds**.  Let's try it with Zinc 1.0.0.
+  
+[![scastie](/resources/img/blog/zinc-1.0-scalatest.gif)](/resources/img/blog/zinc-1.0-scalatest.gif)
+  
+**Zinc 1.0.0** terminates **compilation in only 3 seconds**. For this particular
+example, the class-based dependency in the new version does 7x better
+incremental compilation than the source-based old version.
+  
+This experiment is subject to diverse results depending on the scenarios, but
+the general idea is that we can expect nice speedups in common day-to-day
+situations.  In my experience testing and using Zinc 1.0, I have observed
+speedups of between 2x and 22x.
+  
+### A look at the new algorithm
+  
+In our previous example, we showcase the impact that the new dependency
+detection algorithm has in real-world codebases. However, it may be surprising
+to some users that the previous version was so slow in the first place; 21
+seconds to compile a method addition seems *bad* even for a codebase of so many
+lines of code and such complexity. Why was that the case?
+  
+Zinc 0.13.x has always had an underlying dependency detection operating at the
+source rather than the class level. In practice, this means that the
+invalidation algorithm collected more recompilation candidates than it should:
+all classes defined in a source file that were "affected" by a change would be
+checked for invalidation *transitively*. As it is common to define more than one
+class per source file, the number of source files to recompile grew up quickly
+and resulted in worse incremental compiles.
```

Comment:
I tried rephrasing this paragraph to include some more details, but I'm afraid it has become too detailed. I think it would be great to explain the different dependency contexts and explain why things are recompiled...

> As we said, the big change is class-based dependency tracking, in contrast with the previous source-based dependency tracking. What this means, is that Zinc was tracking dependencies between source files rather than between classes. Since it is very common in Scala to define multiple classes in the same file, let me explain why this change matters so much.
>
> Zinc distinguishes between multiple kinds of dependencies (we call them *contexts*). When `class A` extends `class B`, we say that `A` depends on `B` by inheritance. In this context, any API change (e.g. adding a method, changing a signature, etc.) to `class B` will unconditionally trigger the recompilation of all classes that *transitively depend on it by inheritance*.
>
> Now, consider the following structure:
>
> | Source file | Content                        |
> |------------|-----------------------|
> | A.scala       | class A0<br>class A1 |
> | B.scala       | class B extends A1      |
> | C.scala       | class C extends B       |
> 
> `class B` depends on `class A1`, and `class C` depends on `class B`. Because these dependencies are introduced by inheritance, any API change to `class A1` will trigger the recompilation of `class B` and `class C`. Changing `class A0` should not invalidate any source file (no source files depend on it). However, because Zinc used to track dependencies at the source file level, **any API change to `class A0` would trigger the recompilation of `class B` and `class C`**.
>
> With class-based dependency tracking, Zinc 1.0 is smart enough to understand that nothing needs to be recompiled after the API of `class A0` is changed. This is exactly what happens in the ScalaTest example showed above: We change the API of a class, and all the source files that transitively depend by inheritance on this file are recompiled. With Zinc 1.0, the incremental compiler can figure that no source files need to be recompiled.

Reply:
That's a nice suggestion, thank you @Duhemm :smile:.
  
I will try to incorporate some of the changes you explain, but I will leave out the clarification of the kinds of dependencies Zinc has because I'm afraid to embark on a technical digression of the incremental compiler.
  
The main reason why I left out many of the technical details is because I don't think understanding them is the main goal of this blog post. I was afraid of making it longer than it already is, and I think all these details should be explained in a technical document like https://github.com/sbt/zinc/issues/435.
  
However, I think that illustrating the source-based dependency detection with the table is a great idea! So I'm adding that to the document.

Reply:
Actually, I think it would be nice to include Martin's explanation here. It's not _too_ long, and it's quite educational for anyone who may be interested in this stuff. Doesn't hurt to keep it in here... It's short enough for those who aren't interested in the details to just skip the section.

Reply:
I've actually ended up adding all his explanation, but I've reduced it to less text so that it looks like it's less dense. The only thing I miss is the technical terms ("contexts").